tests/observability/scripts/perform_kiali_test.txt

```diff
@@ -0,0 +1,161 @@
+  if [ -z $KIALI_PORT_FORWARD_PID ]; then
+    echo 'It seems that Kiali port-forward PID is not set up properly'
+    kill $(netstat -ntpl | grep '127.0.0.1:8888' | awk '{ print $7 }' | cut -d'/' -f1)
+  else
+    kill $KIALI_PORT_FORWARD_PID
+  fi
+  exit $exit_code
+}
+
+## Setup port-forwarding for Kiali and register trap
+## for cleaning up on exit
+echo "Setting up port forwarding from 8888 to Kiali's API port"
+setup_kiali_port_forwarding 8888 # TODO: Pick a random unused port
+
+echo "Register clean up task for Kiali port-forward on exit"
+trap cleanup_kiali_port_forwarding EXIT
+
+# Allow port-forwarding to be setup in the background
+sleep 4
+
+
+## Login to Kiali using demo secret and save the token
+echo "Obtain token for accessing Kiali API"
+DEMO_SECRET='YWRtaW46YWRtaW4='
+API_TOKEN=$(curl -H "Authorization: Basic $DEMO_SECRET" $KIALI_LOCAL_URL/authenticate | jq -r '.token')
+if [ -z $API_TOKEN ]; then
+  echo 'Failed to obtain auth token for Kiali'
+  exit 1
+fi
+
+
+# Attaches JWT obtained from Kiali and send
+# API request to Kiali 
+function curl_with_kiali_cred() {
+  local api_path=$1
+  curl -H "Authorization: Bearer $API_TOKEN" $KIALI_LOCAL_URL/$api_path
+}
+
+## Determine Kiali's state
+echo "Checking Kiali state. Make sure it is running"
+KIALI_STATE=$(curl_with_kiali_cred status | jq -r '.status."Kiali state"') 
+if [ $KIALI_STATE != "running" ]; then 
+  echo "kiali is not running. So no point in proceeding further"
+  exit 1; 
+fi
+
+
+## Get ingress host and port
+echo "Determine ingress host and port"
+export INGRESS_HOST={{ .ingressHostCommand }}
+export INGRESS_PORT={{ .ingressPortCommand }}
+export GATEWAY_URL=$INGRESS_HOST:$INGRESS_PORT
+echo "Ingress gateway URL = $GATEWAY_URL"
+
+## Send a few requests to bookinfo application
+echo "Send a few requests to bookinfo application"
+for i in {1..20}; do
+  STATUS_CODE=$(curl -w '%{http_code}' http://$GATEWAY_URL/productpage)
+  if (($STATUS_CODE != 200)); then
+    echo "expected status 200, but got $STATUS_CODE"
+    exit 1
+  fi
+done
+
+## STEP 1:
+## Obtain namespace graph for istio-io-bookinfo namespace
+## It should show services and the pods that back it along
+## with their version
+NAMESPACE_GRAPH=$(curl_with_kiali_cred 'graph?graphType=versionedApp&namespaces=istio-io-bookinfo')
+
+EXPECTED_WORKLOADS=('details-v1' 'istio-ingressgateway' 'productpage-v1' 'ratings-v1' 'reviews-v1' 'reviews-v2' 'reviews-v3')
+WORKLOADS_PATH='.elements.nodes[].data.workload'
+ACTUAL_WORKLOADS=($(echo "${NAMESPACE_GRAPH}" | jq -r $WORKLOADS_PATH))
+EXPECTED_WORKLOADS_STR="${EXPECTED_WORKLOADS[@]}"
+ACTUAL_WORKLOADS_STR="${ACTUAL_WORKLOADS[@]}"
+if [ "$EXPECTED_WORKLOADS_STR" != "$ACTUAL_WORKLOADS_STR" ]; then
+  echo "mismatch in workloads list"
+  echo "expected: $EXPECTED_WORKLOADS_STR"
+  echo "  actual: $ACTUAL_WORKLOADS_STR"
+  exit 1
+fi
+
+## STEP 2:
+## Obtain App graph which aggregates all versions of an app
+## to a single node.
+
+
```

Comment:
Are these steps intentionally left blank?

Reply:
Yes, Each of them involves parsing JSON. At that that I was not even sure what steps should be there. Since much of it would be tested by Kiali team anyways we would test integration with Istio.